### PR TITLE
Render modules and elements from templates

### PR DESCRIPTION
**Pull request by aschempp**
This is a very simple approach that allows to render any module or content element directly in Twig templates without having a database record.

This works perfectly in `fe_page.html.twig`:

```twig
{{ insert_module('navigation', { levelOffset: 1 }) }}
```

it also works to include an existing content element or module, just as the insert tag does.

```twig
{{ insert_module(15) }}
```

### CODE REVIEW

**Comment by m-vo** [comment]
Related https://github.com/contao/contao/pull/3942

**Review by ausi — DISMISSED** [review]

**Review by Toflar — CHANGES_REQUESTED** [review]
Tests...

**Comment by m-vo** [comment]
I think I would prefer if the method would return a controller reference and I could use it analogous to the (upcoming) nested content elements with something like `{{ render(content_element(…)) }}`. 

(I would not care about legacy elements/modules.)

**Comment by aschempp** [comment]
> I think I would prefer if the method would return a controller reference and I could use it analogous to the (upcoming) nested content elements with something like `{{ render(content_element(…)) }}`.

This would have at least two major negative implications which would make this PR useless to me
 1. you cannot use any existing module (almost all core modules) and same goes for any extension that is not updated to fragments (I would guess thats > 80%)
 2. permissions or visibility would not be checked when including (contrary to what the insert tag does)

**Comment by m-vo** [comment]
> permissions or visibility

Which permissions do you mean? Visibility also doesn't make any sense for on-the-fly generated things, no?

**Comment by ausi** [comment]
> 1. you cannot use any existing module

Could we create a proxy that is the other way round? Meaning a proxy fragment controller that renders legacy elements? In this case we could then always return a controller reference.

**Comment by aschempp** [comment]
I don't think that is worth the effort or makes sense. It also would not solve the permission/publishing checks etc.

File: core-bundle/src/Twig/Extension/ContaoExtension.php
Comment (ausi) [inline core-bundle/src/Twig/Extension/ContaoExtension.php]:
```suggestion
                'frontend_module',
                [FragmentRuntime::class, 'renderModule'],
                ['is_safe' => ['html']]
            ),
            new TwigFunction(
                'content_element',
```

**Review by leofeyer — DISMISSED** [review]
The changes look good to me now. I haven't tested them though, so I will wait for another approval from @ausi or @Toflar.

**Review by m-vo — DISMISSED** [review]
Reviewed & manually tested. Works fine! :+1:

**Comment by m-vo** [comment]
One thing, that I found out when testing, is, that providing normally serialized values as an object currently already works:

```twig
{{ content_element('text', {
    headline: {
        unit: 'h1',
        value: 'Inline content element example'
    },
    text: '<p>Here you go!</p>'
}) }}
```

But this is because the value just gets set in the model (as an associative array) and `StringUtil::deserialize()` in the controller then skips and returns the value (that is already an array). If the model would have a type check, this would fail. I wonder if we should handle the case by either never allowing objects/arrays (and adding a `|serialize` filter) or by automatically serializing everything that's not a scalar. :thinking:

**Comment by aschempp** [comment]
I think that's perfectly good to work like that. Why would I want to pass serialized data if I don't need to?

**Comment by m-vo** [comment]
> Why would I want to pass serialized data if I don't need to?

Because `$contentModel->headline` is supposed to be a string, not an array:
https://github.com/contao/contao/blob/03bdd806220338a2be1f3f019f9bf81d6dc0467d/core-bundle/contao/models/ContentModel.php#L25

If there would be a setter, it would have the signature `setHeadline(string $headline)` and would break, same with a getter or real property.

**Comment by ausi** [comment]
> I wonder if we should handle the case by either never allowing objects/arrays (and adding a `|serialize` filter) or by automatically serializing everything that's not a scalar.

I think the latter makes most sense, this way your example code would work as expected 👍

> Why would I want to pass serialized data if I don't need to?

I think there should be no difference (or as little as possible) between the data coming from the database and the data coming from the template directly.

**Review by Toflar — DISMISSED** [review]

**Comment by leofeyer** [comment]
> > I wonder if we should handle the case by either never allowing objects/arrays (and adding a `|serialize` filter) or by automatically serializing everything that's not a scalar.
> 
> I think the latter makes most sense, this way your example code would work as expected 👍

So this means that the pull request needs to be adjusted again?

File: core-bundle/src/Twig/Runtime/FragmentRuntime.php
Comment (ausi) [inline core-bundle/src/Twig/Runtime/FragmentRuntime.php:61]:
```suggestion
            if (null !== $v && !\is_scalar($v)) {
                $v = serialize($v);
            }

            $model->$k = $v;
```

File: core-bundle/tests/Twig/Runtime/FragmentRuntimeTest.php
Comment (ausi) [inline core-bundle/tests/Twig/Runtime/FragmentRuntimeTest.php]:
```suggestion
        $result = $runtime->renderContent('text', ['foo' => 'bar', 'headline' => ['unit' => 'h2', 'value' => 'Test']]);
```

File: core-bundle/tests/Twig/Runtime/FragmentRuntimeTest.php
Comment (ausi) [inline core-bundle/tests/Twig/Runtime/FragmentRuntimeTest.php]:
```suggestion
                    $this->assertSame(['type' => 'text', 'foo' => 'bar', 'headline' => serialize(['unit' => 'h2', 'value' => 'Test'])], $model->row());
```

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**
